### DISCUSSION TITLE
URL & Unicode characters

### DISCUSSION THREAD

**Discussion by van-de-bugger** [jgm/pandoc#9905] [General]
Hi,

I faced a problem related to Unicode (non-ASCII) characters and URLs. Look:

[HTML standard](https://html.spec.whatwg.org/#the-id-attribute) applies very small restrictions on ids: an id must be non-empty and unique, that's all. So, an id of non-ASCII characters is perfectly valid:

    # Section Ψ {#Ψ}

I used a Greek letter as id. Ok, let's how use this as a link destination:

    ... look at the [Section Ψ](#Ψ}. ...

And see how pandoc process it:

    $ cat test.md 
    # Section Ψ {#Ψ}

    ... look at the [Section Ψ](#Ψ). ...

    $ pandoc -f markdown -t html test.md | tee test.html
    <h1 id="Ψ">Section Ψ</h1>
    <p>… look at the <a href="#Ψ">Section Ψ</a>. …</p>

    $ tidy -errors test.html 
    line 1 column 1 - Warning: missing <!DOCTYPE> declaration
    line 1 column 1 - Warning: inserting implicit <body>
    line 1 column 1 - Warning: inserting missing 'title' element
    line 2 column 18 - Warning: <a> escaping malformed URI reference
    Info: Document content looks like HTML5
    Tidy found 4 warnings and 0 errors!

    URIs must be properly escaped, they must not contain unescaped
    characters below U+0021 including the space character and not
    above U+007E. Tidy escapes the URI for you as recommended by
    HTML 4.01 section B.2.1 and XML 1.0 section 4.2.2. Some user agents
    use another algorithm to escape such URIs and some server-sided
    scripts depend on that. If you want to depend on that, you must
    escape the URI on your own. For more information please refer to
    http://www.w3.org/International/O-URL-and-ident.html

    <...>

Let's ignore the first 3 tidy warnings: it is because I generated a HTML fragment, not a standalone file. The fourth warning is my point. pandoc passed the id of the destination to the link's href attribute as-is: `<a href="#Ψ">`, but href is expected to be URL, and URL syntax does not allow non-ASCII characters. Non-ASCII characters in URLs (let's forget about domain part for a moment) must be percent-encoded, so the correct HTML code will be: `<a href="#%CE%A8">`.

I can modify the markdown source to get fully correct HTML:

    ... look at the [Section Ψ](#%CE%A8). ...

But it looks ugly, contradicts to the markdown philosophy that markdown must be human-readable, and unmaintainable.

The same problem arises with Unicode characters in URL paths, e. g.:

    $ echo '[Voilà](https://en.wikipedia.org/wiki/Voilà_(Barbara_Pravi_song))' | pandoc -f markdown -t html
    <p><a href="https://en.wikipedia.org/wiki/Voilà_(Barbara_Pravi_song)">Voilà</a></p>

In such a case not only `à`, but also parenthesis must be percent-encoded. The special encoding rules is used for characters in domain part of an URL.

The problem is not very serious, though, since modern browsers accept non-ASCII characters in URLs: I tested Firefox and Chromium and didn't see any problem with such links.

What do you think? Should pandoc properly encode URLs to generate valid HTML, or pass them as is and hope browsers will do the work?

P. S. I think the problem has been discussed, but I can't find it in issues or in discussions.

**Comment by tarleb** [comment]
Note: these links are perfectly fine in HTML5, so the issue concerns only HTML4 (and EPUB2).

**Reply by jgm** [reply]
At one time we did URI-encode the links.  I have forgotten why we stopped.
From pandoc 1.9 changelog:
```
 * Changes in URI escaping:  Previously the readers escaped URIs by
    converting unicode characters to octets and then percent encoding.
    Now unicode characters are left as they are, and `escapeURI` only
    percent-encodes space characters.  This gives more readable
    URIs, and works well with modern user agents. URIs are no longer unescaped
    at all on conversion to `markdown`, `asciidoc`, `rst`, `org`.
```

**Reply by jgm** [reply]
We could have the HTML4/EPUB2 writer escape the URIs, I suppose.  The fact that we have not been doing this since pandoc 1.9 strongly suggests that the issue is merely academic.

**Comment by jgm** [comment]
I made the change.

**Comment by van-de-bugger** [comment]
One more question related to non-ASCII characters: What does pandoc consider a valid id?

I faced a funny case:

    $ echo "[link](#❦)" | pandoc
   <p><a href="#❦">link</a></p>

You see, `#❦` is recognized as proper link destination, `a` element is generated.

Now let's try to generate a span with such id:

    $ echo "[id]{#❦}" | pandoc
    <p>[id]{#❦}</p>
 
Oops, the construct is not recognized as a span at all. There is a workaround, though:

    $ echo "[id]{id=❦}" | pandoc
    <p><span id="❦">id</span></p>

HTML allows any characters in id, but pandoc can't do the same, since space is used to separate id from other attributes, closing brace terminates the list of attributes, etc. Currently pandoc documentation does not specify what characters are allowed and what are not:

>   Attributes can be set on links and images:
>
>       An inline ![image](foo.jpg){#id .class width=30 height=20px}
>       and a reference ![image][ref] with attributes.

**Reply by jgm** [reply]
Pandoc requires the identifier to start with a letter and continue with alphanumerics or one of `-_:.`.
src/Text/Pandoc/Readers/Markdown.hs l. 638

**Reply by van-de-bugger** [reply]
>   Pandoc requires the identifier to start with a letter and continue with alphanumerics or one of `-_:.`.
    src/Text/Pandoc/Readers/Markdown.hs l. 638

Sorry, I can't read Haskel sources.

But you see, that pandoc is *not* consistent. Sometimes (in a span) it requires the id to start with a letter and so so on, but sometimes (in a link) it does not.

**Reply by tarleb** [reply]
In the case of a link, the pound sign `#` marks a URI fragment, which *doesn't* have to be an ID, and allows more characters.

So as far as I can see, pandoc's Markdown syntax is consistent when it comes to places where an element ID is defined.

**Comment by jidanni** [comment]
https://github.com/htacg/tidy-html5/issues/766
